File: batalha-naval/src/Controle/Jogo.hs
{-# OPTIONS_GHC -Wno-missing-export-lists #-}
module Controle.Jogo where

import Tipos.Base
import Interface.Interface (printTabuleiro)
import Logica.Atirar (atirar)
import Logica.Posicionamento (posicionarNavio, geraCorpo, coordenadaValida)
import System.IO (hFlush, stdout)
import Data.Char (toUpper)

lerOrientacao :: IO Orientacao
lerOrientacao = do
    putStr "Orientacao (H ou V): "
    hFlush stdout
    entrada <- getLine
    case map toUpper entrada of
        "H" -> return H
        "V" -> return V
        _   -> do
            putStrLn "Orientacao invalida! Digite H (horizontal) ou V (vertical)."
            lerOrientacao

lerInt :: String -> IO Int
lerInt prompt = do
    putStr prompt
    hFlush stdout
    entrada <- getLine
    case reads entrada of
        [(n, "")] | n >= 0 && n <= 9 -> return n
        _ -> do
            putStrLn "Valor invalido! Digite um numero entre 0 e 9."
            lerInt prompt

iniciarJogo :: IO ()
iniciarJogo = do
    putStrLn "\n=== JOGADOR 1: Posicione seus navios ==="
    tabP1 <- fasePosicionamento
    putStrLn "\n=== JOGADOR 2: Posicione seus navios ==="
    tabP2 <- fasePosicionamento
    loopJogo tabP1 tabP2 True

fasePosicionamento :: IO Tabuleiro
fasePosicionamento = posicionarNavios tabuleiroVazio naviosPadrao

posicionarNavios :: Tabuleiro -> [Navios] -> IO Tabuleiro
posicionarNavios tab [] = return tab
posicionarNavios tab (navio:resto) = do
    putStrLn "\nSeu tabuleiro atual:"
    printTabuleiro tab tabuleiroVazio 0
    putStrLn $ "\nPosicione o navio " ++ tipo navio ++ " (tamanho " ++ show (tamanho navio) ++ ")"
    linha  <- lerInt "Linha inicial (0-9): "
    coluna <- lerInt "Coluna inicial (0-9): "
    orient <- lerOrientacao
    let coord = (linha, coluna)
    case posicionarNavio tab coord (tamanho navio) orient of
        Just novoTab -> posicionarNavios novoTab resto
        Nothing      -> do
            let corpo = geraCorpo coord (tamanho navio) orient
            if not (all coordenadaValida corpo)
                then putStrLn "Posicao invalida: o navio ficaria fora do tabuleiro!"
                else putStrLn "Posicao invalida: ja existe um navio nessa posicao!"
            posicionarNavios tab (navio:resto)

esconderNavios :: Tabuleiro -> Tabuleiro
esconderNavios = map (map esconderCelula)
    where
        esconderCelula Navio = Agua
        esconderCelula c     = c

loopJogo :: Tabuleiro -> Tabuleiro -> Bool -> IO ()
loopJogo tabP1 tabP2 turnoP1 = do
    putStrLn "\n========================================"
    if turnoP1
        then putStrLn ">>> VEZ DO JOGADOR 1 <<<"
        else putStrLn ">>> VEZ DO JOGADOR 2 <<<"
    putStrLn "   SEU TABULEIRO                         TABULEIRO INIMIGO"
    if turnoP1
        then printTabuleiro tabP1 (esconderNavios tabP2) 0
        else printTabuleiro tabP2 (esconderNavios tabP1) 0
    putStrLn "\nSua vez de atirar!"
    linha  <- lerInt "Digite a LINHA (0-9): "
    coluna <- lerInt "Digite a COLUNA (0-9): "
    let coord = (linha, coluna)
    if turnoP1
        then do
            let novoTabP2 = atirar coord tabP2
            if jogoAcabou novoTabP2
                then putStrLn "\nJOGADOR 1 VENCEU! Todos os navios inimigos foram afundados!"
                else loopJogo tabP1 novoTabP2 False
        else do
            let novoTabP1 = atirar coord tabP1
            if jogoAcabou novoTabP1
                then putStrLn "\nJOGADOR 2 VENCEU! Todos os navios inimigos foram afundados!"
                else loopJogo novoTabP1 tabP2 True

jogoAcabou :: Tabuleiro -> Bool
jogoAcabou tab = not (any (elem Navio) tab)

File: prolog/batalha-naval/controle/jogo.pl
% =============================================================
% Responsável pelo controle da partida do jogo.
% =============================================================

:- module(jogo, [iniciar_jogo/0]).

:- use_module('../tipos/base',            [tabuleiro_vazio/1, navios_padrao/1]).
:- use_module('../interface/interface',   [print_tabuleiro/2]).
:- use_module('../logica/atirar',         [atirar/3]).
:- use_module('../logica/posicionamento', [posicionar_navio/5, gera_corpo/4, coordenada_valida/1]).

% Lê a orientação escolhida pelo usuário
ler_orientacao(Orientacao) :-
    format("Orientacao (H ou V): "),
    read_line_to_string(user_input, Linha),
    string_upper(Linha, LinhaMaiuscula),
    ( LinhaMaiuscula == "H" -> Orientacao = h
    ; LinhaMaiuscula == "V" -> Orientacao = v
    ; format("Orientacao invalida! Digite H (horizontal) ou V (vertical).~n"),
      ler_orientacao(Orientacao)
    ).

% Lê um número inteiro válido entre 0 e 9
ler_int(Prompt, Valor) :-
    format("~w", [Prompt]),
    read_line_to_string(user_input, Linha),
    ( number_string(N, Linha), integer(N), N >= 0, N =< 9 ->
        Valor = N
    ; format("Valor invalido! Digite um numero entre 0 e 9.~n"),
      ler_int(Prompt, Valor)
    ).

% Inicia a partida completa
iniciar_jogo :-
    format("~n=== JOGADOR 1: Posicione seus navios ===~n"),
    fase_posicionamento(TabP1),
    format("~n=== JOGADOR 2: Posicione seus navios ===~n"),
    fase_posicionamento(TabP2),
    loop_jogo(TabP1, TabP2, true).

% Fase de posicionamento dos navios
fase_posicionamento(Tabuleiro) :-
    tabuleiro_vazio(TabuleiroVazio),
    navios_padrao(Navios),
    posicionar_navios(TabuleiroVazio, Navios, Tabuleiro).

% Posiciona todos os navios do jogador
posicionar_navios(Tabuleiro, [], Tabuleiro).
posicionar_navios(Tabuleiro, [navio(Tipo, Tamanho, _, _) | Resto], TabuleiroFinal) :-
    format("~nSeu tabuleiro atual:~n"),
    tabuleiro_vazio(TabuleiroOculto),
    print_tabuleiro(Tabuleiro, TabuleiroOculto),
    format("~nPosicione o navio ~w (tamanho ~w)~n", [Tipo, Tamanho]),
    ler_int("Linha inicial (0-9): ", Linha),
    ler_int("Coluna inicial (0-9): ", Coluna),
    ler_orientacao(Orientacao),
    Coord = coord(Linha, Coluna),
    ( posicionar_navio(Tabuleiro, Coord, Tamanho, Orientacao, NovoTabuleiro) ->
        posicionar_navios(NovoTabuleiro, Resto, TabuleiroFinal)
    ;
        gera_corpo(Coord, Tamanho, Orientacao, Corpo),
        ( forall(member(C, Corpo), coordenada_valida(C)) ->
            format("Posicao invalida: ja existe um navio nessa posicao!~n")
        ;
            format("Posicao invalida: o navio ficaria fora do tabuleiro!~n")
        ),
        posicionar_navios(Tabuleiro, [navio(Tipo, Tamanho, _, _) | Resto], TabuleiroFinal)
    ).

% Versão do tabuleiro sem exibir os navios
esconder_navios([], []).
esconder_navios([Linha | Resto], [LinhaOculta | RestoOculto]) :-
    esconder_linha(Linha, LinhaOculta),
    esconder_navios(Resto, RestoOculto).

esconder_linha([], []).
esconder_linha([navio | Resto], [agua | RestoOculto]) :- !,
    esconder_linha(Resto, RestoOculto).
esconder_linha([Celula | Resto], [Celula | RestoOculto]) :-
    esconder_linha(Resto, RestoOculto).

% Ciclo principal dos turnos da partida
loop_jogo(TabP1, TabP2, TurnoP1) :-
    format("~n========================================~n"),
    ( TurnoP1 -> format(">>> VEZ DO JOGADOR 1 <<<~n")
    ;            format(">>> VEZ DO JOGADOR 2 <<<~n")
    ),
    format("   SEU TABULEIRO                         TABULEIRO INIMIGO~n"),
    ( TurnoP1 ->
        esconder_navios(TabP2, TabP2Oculto),
        print_tabuleiro(TabP1, TabP2Oculto)
    ;
        esconder_navios(TabP1, TabP1Oculto),
        print_tabuleiro(TabP2, TabP1Oculto)
    ),
    format("~nSua vez de atirar!~n"),
    ler_int("Digite a LINHA (0-9): ", Linha),
    ler_int("Digite a COLUNA (0-9): ", Coluna),
    Coord = coord(Linha, Coluna),
    ( TurnoP1 ->
        atirar(Coord, TabP2, NovoTabP2),
        ( jogo_acabou(NovoTabP2) ->
            format("~nJOGADOR 1 VENCEU! Todos os navios inimigos foram afundados!~n")
        ;
            loop_jogo(TabP1, NovoTabP2, false)
        )
    ;
        atirar(Coord, TabP1, NovoTabP1),
        ( jogo_acabou(NovoTabP1) ->
            format("~nJOGADOR 2 VENCEU! Todos os navios inimigos foram afundados!~n")
        ;
            loop_jogo(NovoTabP1, TabP2, true)
        )
    ).

% Verifica se todos os navios foram destruídos para definir fim
jogo_acabou(Tabuleiro) :-
    \+ (member(Linha, Tabuleiro), member(navio, Linha)).
